pkg/services/publicdashboards/service/service.go

```diff
@@ -253,8 +256,48 @@ func (pd *PublicDashboardServiceImpl) buildMetricRequest(ctx context.Context, da
 	// determine safe resolution to query data at
 	safeInterval, safeResolution := pd.getSafeIntervalAndMaxDataPoints(reqDTO, ts)
 	for i := range queries {
+		datasourceType := queries[i].Get("datasource").Get("type").MustString()
+		pd.log.Info(fmt.Sprintf("datasource type %s", datasourceType))
+
 		queries[i].Set("intervalMs", safeInterval)
 		queries[i].Set("maxDataPoints", safeResolution)
+
+		// replace template variables
+		// this is a POC, not intended to get merged
+		// this loop is replacing all occurrences of ${var} and $var inside rawSql and expr
+		// using the default values declared in the original dashboard
+		// we should have different services for each datasource type (like postgres, prometheus, graphite)
+		// and different methods for each query type (textbox, custom, list, multivalue)
+		for _, varObj := range dashboard.Data.Get("templating").Get("list").MustArray() {
+			variable := simplejson.NewFromAny(varObj)
+			var value string
+
+			if variable.Get("multi").MustBool() {
+				valuesObj := variable.Get("current").Get("value").MustArray()
+				var valuesArray []string
+
+				for _, valueObj := range valuesObj {
+					valuesArray = append(valuesArray, valueObj.(string))
+				}
+				value = strings.Join(valuesArray[:], ",")
+			} else {
+				value = variable.Get("current").Get("value").MustString()
+			}
+
+			name := variable.Get("name").MustString()
+
+			// some DS uses rawSQL
+			rawSql := queries[i].Get("rawSql").MustString()
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("${%s}", name), value, -1)
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("$%s", name), value, -1)
+			queries[i].Set("rawSql", rawSql)
+
+			// some DS uses expr
```

Comment:
there is so many "some DS uses x" scenarios, i wonder if there is a generic approach. Could we string replace the whole blob?

Reply:
you need to send the variables (name+value) to the data source in the data source request query model, each data source will need to implement the interpolation

Reply:
but that is going to be hard for external data sources unless move the interpolation code and variable formats into the go sdk